Sources/SmithyJSON/Writer/Writer.swift

```diff
@@ -0,0 +1,164 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import protocol SmithyReadWrite.SmithyWriter
+import enum SmithyReadWrite.Document
+import enum SmithyTimestamps.TimestampFormat
+import struct SmithyTimestamps.TimestampFormatter
+import struct Foundation.Data
+import struct Foundation.Date
+import class Foundation.NSNumber
+
+public final class Writer: SmithyWriter {
+    public typealias NodeInfo = SmithyJSON.NodeInfo
+
+    let nodeInfo: NodeInfo
+    var jsonNode: JSONNode?
+    var children: [Writer] = []
+    weak var parent: Writer?
+
+    public init(nodeInfo: NodeInfo) {
+        self.nodeInfo = nodeInfo
+    }
+
+    public required init(nodeInfo: NodeInfo, parent: Writer? = nil) {
+        self.nodeInfo = nodeInfo
+        self.parent = parent
+    }
+}
+
+public extension Writer {
+
+    subscript(nodeInfo: NodeInfo) -> Writer {
+        self.jsonNode = .object
+        let newChild = Writer(nodeInfo: nodeInfo, parent: self)
+        children.append(newChild)
+        return newChild
+    }
```

Comment:
That check ensures that you don't have two children with the same nodeInfo.

- XML explicitly permits (and uses this for serializing collections) this so the check is not performed on the XML `Writer`.
- `x-www-form-urlencoded` actually does permit this, but AWS does not use it in its implementation, so the check is included on the FormURL `Writer`.
- JSON does not permit children with the same name, so I will add the check there.  (In practice, because almost all the code that subscripts a `Writer` is code-generated from a validated Smithy model, it is unlikely that it would be needed, but I will add it anyway.)